server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -0,0 +1,120 @@
+package org.http4s.server.middleware
+
+import org.http4s.{Headers, Http, Response, Status}
+import cats.data.Kleisli
+import cats.effect.{Clock, Sync}
+import cats.effect.concurrent.Ref
+import scala.concurrent.duration.FiniteDuration
+import cats.implicits._
+import java.util.concurrent.TimeUnit.NANOSECONDS
+import org.http4s.headers.`Retry-After`
+
+/**
+  * Transform a service to reject any calls the go over a given rate.
+  */
+object Throttle {
+
+  sealed abstract class TokenAvailability extends Product with Serializable
+  case class TokenAvailable() extends TokenAvailability
+  case class TokenUnavailable(retryAfter: Option[FiniteDuration]) extends TokenAvailability
```

Comment:
Is this `Option` even `None`?

Reply:
It could be with other `TokenBucket` impls, ie I envisage an "outstanding requests" version where you issue tokens when there are less than n requests currently being executed, in that scenario there is no obvious time-to-full